src/ivc/protogalaxy/mod.rs

```diff
@@ -0,0 +1,211 @@
+            CurveAffine,
+        },
+        plonk::Error as Halo2PlonkError,
+    };
+
+    use crate::{
+        gadgets::ecc::AssignedPoint,
+        main_gate::{AdviceCyclicAssignor, AssignedValue, MainGateConfig, RegionCtx},
+        nifs::protogalaxy,
+        plonk::PlonkInstance,
+        poseidon::ROCircuitTrait,
+        util::ScalarToBase,
+    };
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Error {
+        #[error("Error while assign {annotation}: {err:?}")]
+        Assign {
+            annotation: &'static str,
+            err: Halo2PlonkError,
+        },
+    }
+
+    /// Assigned version of [`crate::plonk::PlonkInstance`]
+    pub struct AssignedPlonkInstance<C: CurveAffine> {
+        W_commitments: Vec<AssignedPoint<C>>,
+        instances: Vec<Vec<AssignedValue<C::Base>>>,
+        challenges: Vec<AssignedValue<C::Base>>,
+    }
+
+    impl<C: CurveAffine> AssignedPlonkInstance<C> {
+        pub fn assign<const T: usize>(
+            region: &mut RegionCtx<C::Base>,
+            main_gate_config: MainGateConfig<T>,
+            pi: PlonkInstance<C>,
+        ) -> Result<Self, Error> {
+            let PlonkInstance {
+                W_commitments,
+                instances,
+                challenges,
+            } = pi;
+
+            let mut assigner = main_gate_config.advice_cycle_assigner();
+
+            let W_commitments = W_commitments
+                .iter()
+                .enumerate()
+                .map(|(i, W_commitment)| {
+                    assigner.assign_next_advice_point(
+                        region,
+                        || format!("W_commitments[{i}]"),
+                        W_commitment,
+                    )
+                })
+                .collect::<Result<Vec<_>, _>>();
+
+            let instances = instances
+                .iter()
+                .map(|instance| {
+                    assigner.assign_all_advice(
+                        region,
+                        || "instance",
+                        instance.iter().map(|i| C::scalar_to_base(i).unwrap()),
+                    )
+                })
+                .collect::<Result<Vec<_>, _>>();
+
+            let challenges = assigner.assign_all_advice(
+                region,
+                || "challenges",
+                challenges.iter().map(|i| C::scalar_to_base(i).unwrap()),
+            );
+
+            let map_err = |err| Error::Assign {
+                annotation: "PlonkInstance",
+                err,
+            };
+
+            Ok(Self {
+                W_commitments: W_commitments.map_err(map_err)?,
+                instances: instances.map_err(map_err)?,
+                challenges: challenges.map_err(map_err)?,
+            })
+        }
+    }
+
+    /// Assigned version of [`crate::nifs::protogalaxy::accumulator::AccumulatorInstance`]
+    pub struct AssignedAccumulatorInstance<C: CurveAffine> {
+        ins: AssignedPlonkInstance<C>,
```

Comment:
So the current plan is to stay with folding of one intance `k=1` for now?

Reply:
No, it's on-circuit representation of https://github.com/snarkify/sirius/blob/31b9f9639ebf8096118fe36e96a48c80a89ce637/src/nifs/protogalaxy/accumulator.rs#L57-L59 this field

At the end of new file in `fold` fn type of argument with incoming instances a slice, no any blockers at this stage for folding of any count of it